feat: initial mobile support

FabianLars opened the pull request:
ref https://github.com/tauri-apps/tauri-action/issues/525 (not closing it)

fu050409 commented:
Any updates on this pull request? Tauri is v2 stable now :D

AbandonedCart commented:
includeAndroid / includeIOS isn’t entirely unheard of. In a lot of binaries, they use flags that are nothing more than `withModule` or `usesModule`

Not sure if it will be any use to you, but I came across an article for how to run automated testing in an action (if that’s even what you’re setting up)

https://medium.com/@insub4067/ios-automating-ios-tests-with-github-actions-and-xctest-19166b6d26d4

fu050409 commented:
Ah...I have to say I can't agree with you more, iOS does cause a lot of trouble (even when built on local machines). Have you ever considered splitting iOS and Android? We can start with a partial solution for Android only, which already solves some of the problems.

FabianLars commented:
nah i'll add ios too but let users test it instead of testing it in this repo. Apple gatekeeps .ipa generation behind a (paid?) developer account even though it's literally just a zip file so we can't test it ourselves here yet.

just need to update the search paths, some initial docs and we're good to go. idk if that will happen today but at least this week.

btw, in case someone didn't know, you can use PRs in your own workflow. in this case by replacing `tauri-apps/tauri-action@v0` with `tauri-apps/tauri-action@mobile` (will break once the branch is deleted so using commit hashes instead of the branch name would prob be a good idea)

AbandonedCart commented:
> nah i'll add ios too but let users test it instead of testing it in this repo. Apple gatekeeps .ipa generation behind a (paid?) developer account even though it's literally just a zip file so we can't test it ourselves here yet.

You only need a paid account for entitlements, which usually implies some sort of distribution. You can generate an IPA and test it locally with Xcode for free, so you'd be fine leaving it to the user. Doing it without Xcode is hacky and convoluted, though.


But, the actual reason I was here is...

Android explicitly disabled clear text traffic and enabling it requires access to the manifest. I have yet to find any sort of command or anything that would change it without manually editing the manifest. That more or less makes a completely automated build out of the question. Are there any plans to support something to handle that (down the road)?

FabianLars commented:
yes, something like this (iirc "patch" files) is planned but mostly unrelated to this repo.
This is why the templates don't have the gen/android & gen/apple folders in the .gitignore files at the moment.

AbandonedCart commented:
Thanks for the info. I had a feeling it may be somewhat unrelated, but also means anyone who needs such edits won't be able to use it. Not your fault, though.

YuenSzeHong commented:
Any updates on the android support? it has been 2 months since last commit

FabianLars commented:
Android should work

YuenSzeHong commented:
> Android should work. check out the git diff of this pr for an example workflow (it's missing signing which is required to be able to install the APK!)

will the signing be included? looking for that to be also in the workflow

FabianLars-crabnebula commented:
the test workflow probably won't but i'll add something to the docs/example then. For now it's explained here https://v2.tauri.app/distribute/sign/android/

YuenSzeHong commented:
```kts
    splits {
        abi {
            isEnable = true
            reset()
            include("armeabi-v7a", "arm64-v8a", "x86", "x86_64")
            isUniversalApk = true
        }
    }
```

do I add this to my `build.gradlew.kts` or handle the split at the workflow level? or is this supported?

YuenSzeHong commented:
```yaml
- platform: 'ubuntu-22.04'
  args: '--apk'
  mobile: true
```
```yaml
- uses: tauri-apps/tauri-action@3c20edcd47f6c6fa6d7bd17d279e1c1b454ff57d
  env:
    GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
  with:
    tagName: v__VERSION__ # the action automatically replaces \_\_VERSION\_\_ with the app version.
    releaseName: 'v__VERSION__'
    releaseBody: 'See the assets to download this version and install.'
    releaseDraft: true
    prerelease: false
    includeAndroid: ${{ matrix.mobile || false }}
    args: ${{ matrix.args }}
```
```
Run tauri-apps/tauri-action@3c20edcd47f6c6fa6d7bd17d279e1c1b454ff57d
  with:
    tagName: v__VERSION__
    releaseName: v__VERSION__
    releaseBody: See the assets to download this version and install.
    releaseDraft: true
    prerelease: false
    includeAndroid: true
    args: --apk
    projectPath: .
    includeDebug: false
    includeRelease: true
    includeIOS: false
    includeUpdaterJson: true
    updaterJsonKeepUniversal: false
  env:
    CARGO_HOME: /home/runner/.cargo
    CARGO_INCREMENTAL: 0
    CARGO_TERM_COLOR: always
    CACHE_ON_FAILURE: false
    JAVA_HOME: /opt/hostedtoolcache/Java_Zulu_jdk/17.0.13-11/x64
    JAVA_HOME_17_X64: /opt/hostedtoolcache/Java_Zulu_jdk/17.0.13-11/x64
    ANDROID_HOME: /usr/local/lib/android/sdk
    ANDROID_SDK_ROOT: /usr/local/lib/android/sdk
    GITHUB_TOKEN: ***
running bun [ 'tauri', 'build', '--apk' ]
$ tauri build --apk
error: unexpected argument '--apk' found

  tip: to pass '--apk' as a value, use '-- --apk'

Usage: bun run tauri build [OPTIONS] [ARGS]...

For more information, try '--help'.
```
is this just me or there is something wrong with my setup?

AbandonedCart commented:
`args: --apk` should be `-- --apk` according to the error

YuenSzeHong commented:
> `args: --apk` should be `-- --apk` according to the error
```
error: unexpected argument '--apk' found

Usage: cargo build [OPTIONS]

For more information, try '--help'.
failed to build app: failed to build app
    Error failed to build app: failed to build app
error: script "tauri" exited with code 1
Error: Command failed with exit code 1: bun tauri build -- --apk
```

wasted my time

YuenSzeHong commented:
> @YuenSzeHong To be honest, Idk what would you like to do. Tauri use `bun tauri android build` to create a android apk bundle instead of using `--apk`, do you mean `android build`?

well, thats what exactly i want to do, i want to do desktop and android build in same matrix and it don't work

FabianLars commented:
You will need 2 Ubuntu entries in the matrix. One for desktop and one for mobile with the extra cli args.

AbandonedCart commented:
> wasted my time, the issue lies on for some reason the includeAndroid is ignored

```
error: unexpected argument '--apk' found

  tip: to pass '--apk' as a value, use '-- --apk'
```

My apologies, I guess? I diagnosed the error, not the entire build.


> You will need 2 Ubuntu entries in the matrix. One for desktop and one for mobile with the extra cli args.

It should be noted somewhere that iOS will only build if `runs-on: macos` is set.

FabianLars commented:
> It should be noted somewhere that iOS will only build if runs-on: macos is set.

![Screenshot_20241122-151624.png](https://github.com/user-attachments/assets/f63b0659-e9b0-4985-a470-55f552f85ce2)

This will of course be added to the readme too once I actually update the readme.

AbandonedCart commented:
It may be easier to flat out say it only runs on macOS runners. I can see someone reading that and being like “well how do I fix that?”

FabianLars commented:
Hmm yeah, the wording is not final and I will keep that in mind, thanks. Just wanted to make it clear that setting it on the wrong runner won't make the workflow fail (though maybe it should).

AbandonedCart commented:
`error: unrecognized subcommand 'ios'` is the error Tauri throws on Linux and that’s easily mistaken for Tauri being broken.

FabianLars commented:
Yeah, but that is not really relevant for us here. Again, at least with the current plan, the action will ignore the includeIOS setting on Linux/windows and therefore just skip the whole build.

The current implementation may not do that yet, it's a draft after all.

AbandonedCart commented:
Apologies if my intention was unclear. I am aware it is a draft, but must have missed where it ignores the command on other platforms. Since I don't know your actual plans until they are published, I will hold any comments until it's no longer a draft. 

Best of luck.

FabianLars commented:
Yeah it's probably not in the code yet, but that's what I wanted to say in the screenshot :)

Anyway, thanks for the feedback so far!

YuenSzeHong commented:
> You will need 2 Ubuntu entries in the matrix. One for desktop and one for mobile with the extra cli args.

i have tried that, but the flow just failed instead of running the android flow with arguments, i assume the arguments for action should be the same as tauri android build?

YuenSzeHong commented:
do i need to pass `--split-per-abi` to get the splits? as i previously passed in --apk it failed

FabianLars commented:
> do i need to pass --split-per-abi to get the splits?

No, that just splits the one apk into four.

----

i _assume_ that you added the args for all or multiple jobs (otherwise the error doesn't make sense) but you have to add it to just the android one. something like this (copied from your workflow above)
```yml
      matrix:
        include:
          - platform: 'macos-latest' # for Arm based macs (M1 and above).
            args: '--target aarch64-apple-darwin'
          - platform: 'macos-latest' # for Intel based macs.
            args: '--target x86_64-apple-darwin'
          - platform: 'macos-latest' # ios
            ios: true
            args: ''
          - platform: 'ubuntu-22.04' # desktop linux
            args: ''
          - platform: 'ubuntu-22.04' # android
            android: true
            args: '--apk'
            mobile: true
          - platform: 'windows-latest'
            args: ''

......

      - uses: tauri-apps/tauri-action@3c20edcd47f6c6fa6d7bd17d279e1c1b454ff57d
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          NDK_HOME: ${{ env.ANDROID_HOME }}/ndk/27.0.11902837
        with:
          tagName: v__VERSION__ # the action automatically replaces \_\_VERSION\_\_ with the app version.
          releaseName: 'v__VERSION__'
          releaseBody: 'See the assets to download this version and install.'
          releaseDraft: true
          prerelease: false
          includeRelease: ${{ !(matrix.ios || matrix.android) }} # disable desktop builds if we're building for mobile
          includeIOS: ${{ matrix.ios || false }}
          includeAndroid: ${{ matrix.android || false }}
          args: ${{ matrix.args }}
```

note that i typed it in github so it may contain syntax errors

YuenSzeHong commented:
```yml
- platform: 'ubuntu-22.04' # desktop linux
  args: ''
- platform: 'ubuntu-22.04' # android
  args: '--apk'
  mobile: true
```
the second one runs tauri build first then tauri android build, causing my error in the first place, so i moved to single Ubuntu entry with `mobile: true` with no arguments

FabianLars commented:
> the second one runs tauri build first then tauri android build, causing my error in the first place, so i moved to single Ubuntu entry with mobile: true with no arguments

This should be taken care of by the includeRelease line in my last message

FabianLars commented:
note that all the workflow configs from above are now invalid, i changed the inputs a bit.

i'll worry about testing ios later and added a note to the input saying that it's considered unstable for now.

AbandonedCart commented:
> so I am wondering how can i pass arguments to `tauri android build`

Here is how I did it:
```
  strategy:
      fail-fast: false
      matrix:
        include:
          - platform: 'macos-latest' # for Arm based macs (M1 and above).
            args: '--target aarch64-apple-darwin'
          - platform: 'macos-latest' # for Intel based macs.
            args: '--target x86_64-apple-darwin'
          - platform: 'ubuntu-22.04'
            args: ''
          - platform: 'windows-latest'
            args: ''
          - platform: 'ubuntu-latest'
            args: '--target aarch64' # arm64-v8a
            android: true
            arch: 'arm64-v8a'
          - platform: 'ubuntu-latest'
            args: '--target armv7' # armeabi-v7a
            android: true
            arch: 'armeabi-v7a'
          - platform: 'ubuntu-latest'
            args: '--target x86_64'
            android: true
            arch: 'x86_64'
          - platform: 'ubuntu-latest'
            args: '--target i686' # x86
            android: true
            arch: 'x86'
```

I was then able to use the `matrix.arch` in later conditions. Bear in mind, every build is labeled “universal” for android.

FabianLars commented:
> Disabling release on mobile actually caused the mobile builds to fail in my tests.

do you have workflow logs for that? the changed workflow in this pr only bulds debug builds and as far as i can see without issues.

AbandonedCart commented:
I’d have to add it back and run a build. I didn’t keep the run.

AbandonedCart commented:
I think the issue is that when you opt out of the release, you need to specifically opt into the debug. The error that comes up is that no artifacts were found.

ReiLoko4 commented:
> note that all the workflow configs from above are now invalid, i changed the inputs a bit.
> 
> i'll worry about testing ios later and added a note to the input saying that it's considered unstable for now.

So, the examples that were given here dont work anymore? 

I'm sorry to say that but I would be really happy to have a readme for the mobile version, just putting name args should be enough for dudes lime me :)

Thanks for the work btw.

FabianLars commented on README.md:
```diff
@@ -98,16 +98,18 @@ These inputs are only used if your GitHub repository does not contain an existin
+| `includeDebug`             |  false   | whether to include a debug build or not                                                                                                                                                                                                                                                                                                                               ...
+| `includeRelease`           |  false   | whether to include a release build or not                                                                                                                                                                                                                                                                                                                             ...
+| `mobile`                   |  false   | Toggle between desktop and mobile builds, use includeDebug and includeRelease to switch between release and debug builds. Set it to `true` to enable iOS builds on macOS runners and Android builds on Ubuntu runners. Set it to `android` to enable Android builds on all runners. `ios` behaves the same as `true` and only works on macOS runners. Note that this  ...
+| `includeIOS`               |  false   | whether to include an iOS release build or not. macOS only (will be ignored on Linux/Windows).                                                                                                                                                                                                                                                                        ...
```
```suggestion
```
leftover

FabianLars commented:
> So, the examples that were given here dont work anymore?

correct. checkout the files of this pr though: 
https://github.com/tauri-apps/tauri-action/pull/924/files#diff-9ccfc6831832281d3954ef7726f4e6f48623922b4054660913b9970a47774558R227
https://github.com/tauri-apps/tauri-action/pull/924/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R106
(for me it doesn't link to the line perfectly, just scroll up a tiny tiny bit to the `mobile:` property.

AbandonedCart commented:
Has this been abandoned?

Having switched to it when it was moved out of draft, I kept waiting to drop it again with the assumption it was going to be merged or at least rebased.

FabianLars commented:
somewhat yes. i'll add it to next week's tasks

AbandonedCart commented:
> somewhat yes. i'll add it to next week's tasks

I’m not sure how that was intended to be taken, but it seems like the safer bet is not to set high expectations.

ReiLoko4 commented:
Didn't work here even in `v0`

![image](https://github.com/user-attachments/assets/5bc27d1c-203c-463f-9676-cf015689d2d3)


my workflow and my tauri info

`publish.yml`
```yml
name: publish

on:
  push:
    branches:
      - main

permissions:
  contents: write
  packages: write
  pull-requests: write

jobs:
  changelog:
    name: Generate Changelog
    runs-on: ubuntu-latest
    outputs:
      tag: ${{ steps.changelog.outputs.tag }}
      skipped: ${{ steps.changelog.outputs.skipped }}
      version: ${{ steps.changelog.outputs.version }}
      changelog: ${{ steps.changelog.outputs.changelog }}
    steps:
      - name: Checkout
        uses: actions/checkout@v4
      - uses: oven-sh/setup-bun@v2

      - name: Install front-end dependencies
        run: bun install

      - name: setup node
        uses: actions/setup-node@v4
        with:
          node-version: 20

      - name: Generate Changelog
        id: changelog
        uses: TriPSs/conventional-changelog-action@v5.2.1
        with:
          github-token: ${{ secrets.GITHUB_TOKEN }}
          git-user-name: "MangaYouKnow Bot"
          git-user-email: "admin@myk.com"
          config-file-path: ./.github/.config.cjs
          version-file: "./src-tauri/tauri.conf.json"
          skip-version-file: false
          skip-git-pull: true
          pre-release: false

      - run: |
          rm -rf node_modules
          rm package.json
  release:
    name: Release
    needs: changelog
    if: ${{ needs.changelog.outputs.skipped != 'true' }}
    strategy:
      fail-fast: false
      matrix:
        include:
          - platform: "macos-latest" # Arm based macs
            args: "--target aarch64-apple-darwin"
          - platform: "macos-latest" # Intel based macs
            args: "--target x86_64-apple-darwin"
          - platform: "ubuntu-22.04"
            args: ""
          - platform: "windows-latest"
            args: ""

    runs-on: ${{ matrix.platform }}
    steps:
      - name: Checkout
        uses: actions/checkout@v4
      - uses: oven-sh/setup-bun@v2

      - name: Setup node
        uses: actions/setup-node@v4
        with:
          node-version: 22

      - name: Install Rust stable
        uses: dtolnay/rust-toolchain@stable
        with:
          targets: ${{ matrix.platform == 'macos-latest' && 'aarch64-apple-darwin,x86_64-apple-darwin' || '' }}

      - name: Install dependencies (ubuntu only)
        if: matrix.platform == 'ubuntu-22.04'
        run: |
          sudo apt-get update
          sudo apt-get install -y libwebkit2gtk-4.1-dev libappindicator3-dev librsvg2-dev patchelf

      - name: Install front-end dependencies
        run: bun install

      - name: Publish Release
        uses: tauri-apps/tauri-action@v0
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          TAURI_SIGNING_PRIVATE_KEY: ${{ secrets.TAURI_PRIVATE_KEY }}
          TAURI_SIGNING_PRIVATE_KEY_PASSWORD: ${{ secrets.TAURI_KEY_PASSWORD }}
        with:
          tagName: v__VERSION__
          releaseName: v__VERSION__
          appVersion: __VERSION__
          releaseBody: ${{ needs.changelog.outputs.changelog }}
          includeUpdaterJson: true
          mobile: true
          args: ${{ matrix.args }}

      - name: Update Cargo.toml
        run: bash myk u ${{ needs.changelog.outputs.version }}
```

`tauri info`

```
[✔] Environment
    - OS: Windows 10.0.26100 x86_64 (X64)
    ✔ WebView2: 132.0.2957.140
    ✔ MSVC: Visual Studio Build Tools 2022
    ✔ rustc: 1.84.1 (e71f9a9a9 2025-01-27)
    ✔ cargo: 1.84.1 (66221abde 2024-11-19)
    ✔ rustup: 1.27.1 (54dd3d00f 2024-04-24)
    ✔ Rust toolchain: stable-x86_64-pc-windows-msvc (default)
    - node: 22.13.1
    - npm: 10.9.2
    - bun: 1.2.1

[-] Packages
    - tauri 🦀: 2.2.5
    - tauri-build 🦀: 2.0.5
    - wry 🦀: 0.48.0
    - tao 🦀: 0.31.1
    - @tauri-apps/api : 2.2.0
    - @tauri-apps/cli : 2.2.4 (outdated, latest: 2.2.7)

[-] Plugins
    - tauri-plugin-dialog 🦀: 2.2.0
    - @tauri-apps/plugin-dialog : 2.2.0
    - tauri-plugin-shell 🦀: 2.2.0
    - @tauri-apps/plugin-shell : 2.2.0
    - tauri-plugin-log 🦀: 2.2.0
    - @tauri-apps/plugin-log : 2.0.0 (outdated, latest: 2.2.1)
    - tauri-plugin-global-shortcut 🦀: 2.2.0
    - @tauri-apps/plugin-global-shortcut : 2.2.0
    - tauri-plugin-single-instance 🦀: 2.2.1
    - @tauri-apps/plugin-single-instance : not installed!
    - tauri-plugin-window-state 🦀: 2.2.0
    - @tauri-apps/plugin-window-state : 2.2.0 (outdated, latest: 2.2.1)
    - tauri-plugin-sql 🦀: 2.2.0
    - @tauri-apps/plugin-sql : 2.2.0
    - tauri-plugin-fs 🦀: 2.2.0
    - @tauri-apps/plugin-fs : 2.2.0
    - tauri-plugin-http 🦀: 2.2.0
    - @tauri-apps/plugin-http : 2.0.1 (outdated, latest: 2.3.0)
    - tauri-plugin-store 🦀: 2.2.0
    - @tauri-apps/plugin-store : 2.1.0 (outdated, latest: 2.2.0)
    - tauri-plugin-autostart 🦀: 2.2.0
    - @tauri-apps/plugin-autostart : 2.2.0
    - tauri-plugin-notification 🦀: 2.2.0
    - @tauri-apps/plugin-notification : 2.0.0 (outdated, latest: 2.2.1)
    - tauri-plugin-process 🦀: 2.2.0
    - @tauri-apps/plugin-process : 2.2.0
    - tauri-plugin-cli 🦀: 2.2.0
    - @tauri-apps/plugin-cli : 2.2.0
    - tauri-plugin-os 🦀: 2.2.0
    - @tauri-apps/plugin-os : 2.2.0
    - tauri-plugin-updater 🦀: 2.3.1
    - @tauri-apps/plugin-updater : 2.0.0 (outdated, latest: 2.5.0)

[-] App
    - build-type: bundle
    - CSP: http://asset.localhost
    - frontendDist: ../build
    - devUrl: http://localhost:5174/
    - framework: Svelte
    - bundler: Vite
 ```

iWangJiaxiang commented:
Is this ready to merge?

will3942 commented:
@FabianLars FYI - this works well for us for both Android and iOS on this branch with a few fixes (and a sync with dev) https://github.com/will3942/tauri-action/tree/mobile

Fix commits:
- https://github.com/tauri-apps/tauri-action/commit/25c05b84000f64f72d71ad6246bacdda27698b16
- https://github.com/tauri-apps/tauri-action/commit/47f60b859be1377b599d5bbf6111427131812d73
